sdk/keyvault/Azure.Security.KeyVault.Certificates/src/PemReader.cs

```diff
@@ -0,0 +1,191 @@
+
+        private static MethodInfo s_importPkcs8PrivateKeyMethod;
+        private static MethodInfo s_copyWithPrivateKeyMethod;
+
+        /// <summary>
+        /// Loads an <see cref="X509Certificate2"/> from PEM data.
+        /// </summary>
+        /// <param name="data">The PEM data to parse.</param>
+        /// <param name="cer">Optional public certificate data if not defined within the PEM data.</param>
+        /// <param name="allowCertificateOnly">Whether to create an <see cref="X509Certificate2"/> if no private key is read.</param>
+        /// <returns>An <see cref="X509Certificate2"/> loaded from the PEM data.</returns>
+        /// <exception cref="CryptographicException">A cryptographic exception occurred when trying to create the <see cref="X509Certificate2"/>.</exception>
+        /// <exception cref="InvalidOperationException"><paramref name="cer"/> is null and no CERTIFICATE field is defined in PEM, or no PRIVATE KEY is defined in PEM.</exception>
+        /// <exception cref="PlatformNotSupportedException">Creating a <see cref="X509Certificate2"/> from PEM data is not supported on the current platform.</exception>
+        public static X509Certificate2 LoadCertificate(ReadOnlySpan<char> data, byte[] cer = null, bool allowCertificateOnly = false)
+        {
+            byte[] priv = null;
+
+            while (TryRead(data, out PemField field))
+            {
+                if (field.Label.Equals("CERTIFICATE".AsSpan(), StringComparison.Ordinal) && cer is null)
+                {
+                    cer = field.FromBase64Data();
+                    data = data.Slice(field.Start + field.Length);
+                }
+                else if (field.Label.Equals("PRIVATE KEY".AsSpan(), StringComparison.Ordinal))
+                {
+                    priv = field.FromBase64Data();
+                }
+
+                int offset = field.Start + field.Length;
+                if (offset >= data.Length)
+                {
+                    break;
+                }
+
+                data = data.Slice(offset);
+            }
+
+            if (cer is null)
+            {
+                throw new InvalidOperationException("Missing public key");
+            }
+
+            if (priv is null)
+            {
+                if (allowCertificateOnly)
+                {
+                    return new X509Certificate2(cer);
+                }
+
+                throw new InvalidOperationException("Missing private key");
+            }
+
+            if (s_importPkcs8PrivateKeyMethod is null)
+            {
+                // ImportPkcs8PrivateKey was added in .NET Core 3.0 and is only present on Core. We will fall back to a lightweight decoder if this method is missing from the current runtime.
+                s_importPkcs8PrivateKeyMethod = typeof(RSA).GetMethod("ImportPkcs8PrivateKey", BindingFlags.Instance | BindingFlags.Public, null, new[] { typeof(ReadOnlySpan<byte>), typeof(int).MakeByRefType() }, null);
```

Comment:
Would we try to run this code on every `LoadCertificate` call if the `ImportPkcs8PrivateKey` is not present?

Reply:
Yes, but the call is so inconsequential I'm not sure it's worth the overhead (say, an extra `bool`) of keeping track. Feel any different?

Reply:
I don't think it's inconsequential. I expect method binding to be way more expensive than a bool.

It's at minimum a dictionary lookup by name.